File: tests/class_with_dependencies/LoginUseCase.swift
protocol LoginUseCase {
    func callAsFunction(credentials: Credentials) async -> Result<User, LoginError>
}

class LoginUseCaseImpl: LoginUseCase {
    private let authSession: AuthSession
    private let userRepository: UserRepository
    private let passwordValidator: PasswordValidator

    init(authSession: AuthSession,
         userRepository: UserRepository,
         passwordValidator: PasswordValidator) {
        self.authSession = authSession
        self.userRepository = userRepository
        self.passwordValidator = passwordValidator
    }

    func callAsFunction(credentials: Credentials) async -> Result<User, LoginError> {
        do {
            guard passwordValidator.validate(credentials.password) else {
                throw LoginError.invalidPassword("Password does not meet the required criteria")
            }

            let token = try await performLogin(credentials: credentials)
            let user = try await fetchUser(withToken: token)

            return .success(user)
        } catch let error as LoginError {
            return .failure(error)
        } catch {
            return .failure(.unexpected(error.localizedDescription))
        }
    }

    private func performLogin(credentials: Credentials) async throws -> String {
        switch await authSession.login(credentials: credentials) {
        case .success(let token):
            return token
        case .failure(let error):
            throw LoginError.authenticationFailed(error.localizedDescription)
        }
    }

    private func fetchUser(withToken token: String) async throws -> User {
        switch await userRepository.getUser(byId: token) {
        case .success(let user):
            return user
        case .failure(let error):
            throw LoginError.userNotFound(error.localizedDescription)
        }
    }
}

File: tests/class_with_dependencies/login_usecase.dart
import 'dart:async';

enum LoginErrorType { invalidPassword, authenticationFailed, userNotFound, unexpected }

class LoginError implements Exception {
  final LoginErrorType type;
  final String message;

  LoginError(this.type, this.message);
}

abstract class LoginUseCase {
  Future<User> execute(Credentials credentials);
}

class LoginUseCaseImpl implements LoginUseCase {
  final AuthSession authSession;
  final UserRepository userRepository;
  final PasswordValidator passwordValidator;

  LoginUseCaseImpl(this.authSession, this.userRepository, this.passwordValidator);

  @override
  Future<User> execute(Credentials credentials) async {
    try {
      if (!passwordValidator.validate(credentials.password)) {
        throw LoginError(LoginErrorType.invalidPassword, "Password does not meet the required criteria");
      }

      final token = await performLogin(credentials);
      final user = await fetchUser(token);

      return user;
    } on LoginError {
      rethrow;
    } catch (e) {
      throw LoginError(LoginErrorType.unexpected, e.toString());
    }
  }

  Future<String> performLogin(Credentials credentials) async {
    try {
      return await authSession.login(credentials);
    } catch (e) {
      throw LoginError(LoginErrorType.authenticationFailed, e.toString());
    }
  }

  Future<User> fetchUser(String token) async {
    try {
      return await userRepository.getUser(token);
    } catch (e) {
      throw LoginError(LoginErrorType.userNotFound, e.toString());
    }
  }
}